internal/incident/incidents.go

```diff
@@ -132,3 +136,48 @@ func GetCurrentIncidents() map[int64]*Incident {
 	}
 	return m
 }
+
+// ProcessEvent from an event.Event.
+//
+// This function first gets this Event's object.Object and its incident.Incident. Then, after performing some safety
+// checks, it calls the Incident.ProcessEvent method.
+//
+// The returned error might be wrapped around ErrSuperfluousStateChange.
+func ProcessEvent(
+	ctx context.Context,
+	db *icingadb.DB,
+	logs *logging.Logging,
+	runtimeConfig *config.RuntimeConfig,
+	ev *event.Event,
+) error {
+	obj, err := object.FromEvent(ctx, db, ev)
+	if err != nil {
+		return fmt.Errorf("cannot sync event object: %w", err)
+	}
+
+	createIncident := ev.Severity != event.SeverityNone && ev.Severity != event.SeverityOK
+	currentIncident, created, err := GetCurrent(
+		ctx,
+		db,
+		obj,
+		logs.GetChildLogger("incident"),
+		runtimeConfig,
+		createIncident)
+	if err != nil {
+		return fmt.Errorf("cannot get current incident for %q: %w", obj.DisplayName(), err)
+	}
+
+	if currentIncident == nil {
+		switch {
+		case ev.Type == event.TypeAcknowledgement:
+			return fmt.Errorf("%q does not have an active incident, ignoring acknowledgement event from source %d",
+				obj.DisplayName(), ev.SourceId)
+		case ev.Severity != event.SeverityOK:
+			return fmt.Errorf("cannot process event with a non-OK state without a known incident")
+		default:
+			return fmt.Errorf("%w: ok state event from source %d", ErrSuperfluousStateChange, ev.SourceId)
```

Comment:
A similar example is at least shown in the Go blog post [Working with Errors in Go 1.13](https://go.dev/blog/go1.13-errors) within the _Wrapping errors with %w_ section.

I kind of like this approach as it is short on code, but does not rely on some magic effects. Otherwise, implementing a custom error type feels like creating boilerplate code.